### DISCUSSION TITLE
Copilot suggesting lines with different indentation space amounts than editor settings (Jetbrains)

### DISCUSSION THREAD

**Discussion by 7MinutesDead-Git** [community/community#11724] [Copilot Conversations]
Every now and then Copilot will start predicting what I want to write next, but its suggestions will not match the indent settings within Jetbrains IDEs (or at least in this case, WebStorm while writing javascript, since I haven't tested this in the other IDEs).

Most of the time it will suggest code lines indented by 2 spaces, although I have my editor settings set to 4 spaces for javascript.

I *think* it starts properly suggesting 4 spaces indents as long as there is a leading line before it that I have entered first in that scope. But if it's the first line suggested in a new scope / code block, the indentation doesn't always match. I haven't verified this as the exact reproducible behavior yet though, just that it has been occurring sporadically.

Is there a workaround for this?

**Comment by Heath123** [comment]
I'm having a similar issue, in my case the ghost text shows as 2 spaces as it should but the completion gets set to 4, maybe the IDE is trying to auto-indent the code that was already indented?
https://user-images.githubusercontent.com/13787163/170359826-6eabb5ef-de3b-40e1-aa9e-fca2d0232971.mp4

**Reply by OfirBotDev** [reply]
Same here, On the press of enter it adds one to the expected ident, if deleted and click 'tab' it will only go one space.. 
This is new issue for me.. been working with co pilot for some time and did not faced this so far..

**Reply by vidanovic** [reply]
Same here. Quite annoying and still not fixed.

**Reply by fishiu** [reply]
Same here. Quite annoying and still not fixed.

**Reply by minhperry** [reply]
Same thing for HTML/TS/SASS for me

**Reply by zhaocc1106** [reply]
Same for clion. I'm very angry about that.

**Comment by hawkkiller** [comment]
This issue is still relevant

**Comment by kamal-hakim** [comment]
Seems like it's still not fixed.

**Comment by Magorx** [comment]
Seems like it's still not fixed.

**Comment by 7MinutesDead-Git** [comment]
It's now been a year and a half. At this point I'd love to just fix this shit myself

**Comment by mikeatlas** [comment]
Currently seeing tabs inserted via Copilot completions in a Go project that is spaces.

**Comment by vasyan** [comment]
Same problem. Moreover, it kinda conflicts with Webstorm's automatic indent detection ( it doesn't care if it is turned off ) and funked up my formatting here and there.

**Comment by Wizzerinus** [comment]
still relevant. it's particularly nasty in languages like yaml which really do care about the indents

**Comment by UmeshKonduru** [comment]
Still relevant

**Comment by tajnymag** [comment]
Like Wizzerinus said, yaml files are really nasty. It's really hard to write OpenAPI specs, for example, as I have to correct indentations for each autocompleted chunk otherwise the document is invalid.

**Comment by mikeatlas** [comment]
seems to me like this is something that could at least be fixed on the client side relatively easily without having to make it work in the upstream LLM interaction 🤔 just need to adjust output formatting before rendering perhaps?

**Comment by tajnymag** [comment]
> seems to me like this is something that could at least be fixed on the client side relatively easily without having to make it work in the upstream LLM interaction 🤔 just need to adjust output formatting before rendering perhaps?

Sure. However, it still needs to be done in the plugin, not in the IDE. The issue does fall under GitHub/Microsoft, not JetBrains.

**Comment by arthurwolf** [comment]
Having this problem too (vscode)

**Comment by bvolpato** [comment]
For vscode (haven't tried JetBrains), change `indentationMode` to `true` for Python, but it may cause suggestions to come line by line -- so pick your poison, I guess.

**Reply by arthurwolf** [reply]
is this really python-specific, any way to use a similar fix for typescript you think?

**Comment by juasbhu28** [comment]
Thats horrible to try to fix the format in each sentence. :'( Someone please fix this!

**Comment by danieljimenez-mycase** [comment]
Something that helped for jetbrains (using rubymine) was changing editor>code style> language(ruby, javascript, typescript) to have the same indentation as my files.

**Comment by wang818** [comment]
me too

**Comment by jalanubha** [comment]
Still having the indentation problem with VS Code and Copilot! Any solutions?

**Comment by toddsutt** [comment]
I have this issue in RubyMine and Copilot. Please fix this

**Reply by toddsutt** [reply]
I decided to turn off copilot suggestions for html, I don't have this problem as much in ruby files

**Comment by d4r5c0d3** [comment]
I have found something that worked for my case. 
Which was that I was trying to use copilot with language that wants tabs instead of spaces.

Open the Command Palette with `Ctrl + Shift + P` (or `Cmd + Shift + P` on Mac).
Type `Preferences: Open Settings (JSON)` and select it.
In the settings file, add:

```
json
"[gdscript]": {
    "editor.insertSpaces": false,
    "editor.tabSize": 4
}
```

insertSpaces: false 
ensures tabs are used instead of spaces.

tabSize: 4 
sets the width of a tab to 4 spaces

**Comment by oomek** [comment]
Github Copilot chat window is still ignoring the following:
```
"github.copilot.chat.codeGeneration.instructions": [ { "text": "Always use tabs for indentation, not spaces."  }  ],
```
I get tabs replaced with spaces every time.

**Comment by bairesearch** [comment]
Github Copilot need to provide a professional solution for tab indentation support for their chat assistant.

None of these settings.json options work for python (assume the entire codebase is tab delimited);
```
{
	"editor.detectIndentation": false,
	"editor.insertSpaces": false,
	"github.copilot.chat.codeGeneration.instructions": [ { "text": "Always use tabs for indentation, not spaces."  } ],
	"[python]": {
		"editor.insertSpaces": false,
	}
}
```

**Comment by LC43** [comment]
sigh, also me. any code is spaces, so its hard to see what are the real changes and whats just tabs replaced with spaces

**Comment by Mr-Smarty** [comment]
This is still an issue, also in VSCode. Entire files are changed every time converting tab indentations to spaces and it's impossible to see where the real changes are.
